modules/srv6/datapath_headend.c

```diff
@@ -0,0 +1,203 @@
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) 2025 Olivier Gournet
+
+#include <gr_datapath.h>
+#include <gr_eth.h>
+#include <gr_fib6.h>
+#include <gr_graph.h>
+#include <gr_ip4_datapath.h>
+#include <gr_ip6_control.h>
+#include <gr_ip6_datapath.h>
+#include <gr_log.h>
+#include <gr_mbuf.h>
+#include <gr_srv6.h>
+#include <gr_srv6_api.h>
+#include <gr_trace.h>
+#include <gr_vec.h>
+
+#include <rte_byteorder.h>
+#include <rte_ether.h>
+#include <rte_graph_worker.h>
+#include <rte_ip6.h>
+
+#include <netinet/in.h>
+
+//
+// srv6 source node. encapsulate traffic
+//
+
+enum {
+	IP6_OUTPUT = 0,
+	INVALID,
+	NO_ROUTE,
+	NO_HEADROOM,
+	EDGE_COUNT,
+};
+
+struct trace_srv6_data {
+	union {
+		struct ip4_net dest4;
+		struct ip6_net dest6;
+	};
+	bool is_dest6;
+};
+
+static int trace_srv6_format(char *buf, size_t len, const void *data, size_t /*data_len*/) {
+	const struct trace_srv6_data *t = data;
+	if (t->is_dest6)
+		return snprintf(buf, len, "match=" IP6_F "/%hhu", &t->dest6.ip, t->dest6.prefixlen);
+	else
+		return snprintf(buf, len, "match=" IP4_F "/%hhu", &t->dest4.ip, t->dest4.prefixlen);
+}
+
+// called from 'ip6_output' or 'ip_output' node
+static uint16_t srv6_headend_process(
+	struct rte_graph *graph,
+	struct rte_node *node,
+	void **objs,
+	uint16_t nb_objs
+) {
+	struct rte_ipv6_hdr *inner_ip6 = NULL, *outer_ip6;
+	struct rte_ipv4_hdr *inner_ip4 = NULL;
+	struct srv6_policy_data **d_list, *d;
+	struct rte_ipv6_routing_ext *srh;
+	const struct nexthop *nh;
+	struct trace_srv6_data *t = NULL;
+	struct rte_mbuf *m;
+	rte_edge_t edge;
+	uint32_t hdrlen, k, plen;
+	int proto, reduc;
+
+	for (uint16_t i = 0; i < nb_objs; i++) {
+		m = objs[i];
+
+		if (gr_mbuf_is_traced(m))
+			t = gr_mbuf_trace_add(m, node, sizeof(*t));
+
+		// dpdk mbuf packet_type MUST be correctly filled
+		if (m->packet_type & RTE_PTYPE_L3_IPV4) {
```

Comment:
This is a bit fragile. The mbuf packet_type field is dependent on DPDK driver implementations. Maybe you should enforce `m->packet_type |= RTE_PTYPE_L3_IPV4` and `m->packet_type |= RTE_PTYPE_L3_IPV6` from `ip_output` and `ip6_output` to avoid any surprises?